packages/peregrine/lib/talons/CheckoutPage/useCheckoutPage.js

```diff
@@ -142,43 +143,54 @@ export const useCheckoutPage = props => {
     }, [checkoutStep, setCheckoutStep]);
 
     const handlePlaceOrder = useCallback(async () => {
-        try {
-            await getOrderDetails({
-                variables: {
-                    cartId
-                }
-            });
-            await placeOrder({
-                variables: {
-                    cartId
-                }
-            });
-
-            await removeCart();
-
-            await clearCartDataFromCache(apolloClient);
-
-            await createCart({
-                fetchCartId
-            });
-        } catch (err) {
-            console.error(
-                'An error occurred during when placing the order',
-                err
-            );
-            setReviewOrderButtonClicked(false);
-            setCheckoutStep(CHECKOUT_STEP.PAYMENT);
-            // TODO: Delete nonce? The nonce might be expired and why the order
-            // failed. If we delete it the payment info section will render as
-            // if it was not filled, thus prompting the user to enter new info.
+        // Fetch order details and then use an effect to actually place the
+        // order. If/when Apollo returns promises for invokers from useLazyQuery
+        // we can just await this function and then perform the rest of order
+        // placement.
+        getOrderDetails({
+            variables: {
+                cartId
+            }
+        });
+    }, [cartId, getOrderDetails]);
+
+    useEffect(() => {
+        async function placeOrderAndCleanup() {
+            try {
+                await placeOrder({
+                    variables: {
+                        cartId
+                    }
+                });
+
+                await removeCart();
+
+                await clearCartDataFromCache(apolloClient);
+
+                await createCart({
+                    fetchCartId
+                });
+            } catch (err) {
+                console.error(
+                    'An error occurred during when placing the order',
+                    err
+                );
+                setReviewOrderButtonClicked(false);
+                setCheckoutStep(CHECKOUT_STEP.PAYMENT);
+            }
+        }
+
+        if (orderDetailsData && !placeOrderCalled) {
+            placeOrderAndCleanup();
```

Comment:
Should this not be `await placeOrderAndCleanup()`?

Reply:
> Should this not be `await placeOrderAndCleanup()`?

The function provided to `useEffect` isn't an `async` function, so we couldn't `await` this call.

Reply:
Nope - this is the pattern for writing effects that contain async code. You can't have a top level `await` because the callback passed to `useEffect` cannot be asynchronous itself.